core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -403,27 +411,49 @@ class LogManager(logDirs: Seq[File],
    *  Start the background threads to flush logs and do log cleanup
    */
   def startup(topicNames: Set[String]): Unit = {
-    startupWithConfigOverrides(fetchTopicConfigOverrides(topicNames))
+    // ensure consistency between default config and overrides
+    val defaultConfig = currentDefaultConfig
+    startupWithConfigOverrides(defaultConfig, fetchTopicConfigOverrides(defaultConfig, topicNames))
   }
 
   // visible for testing
-  private[log] def fetchTopicConfigOverrides(topicNames: Set[String]): Map[String, LogConfig] = {
+  @nowarn("cat=deprecation")
+  private[log] def fetchTopicConfigOverrides(defaultConfig: LogConfig, topicNames: Set[String]): Map[String, LogConfig] = {
     val topicConfigOverrides = mutable.Map[String, LogConfig]()
-    val defaultProps = currentDefaultConfig.originals()
+    val defaultProps = defaultConfig.originals()
     topicNames.foreach { topicName =>
-      val overrides = configRepository.topicConfig(topicName)
+      var overrides = configRepository.topicConfig(topicName)
```

Comment:
It seems that we get a copy of the config here so I suppose that we could directly mutate it below instead of making a copy again. What do you think?

Reply:
Yes, I thought of that, but we don't specify that the result of `topicConfig` is mutable and we often return unmodifiable collections. Even as it is, we don't do the copy in the common case, so that's why I went with it.

Reply:
For my own education, could `Properties` by unmodifiable somehow? 

I pointed that out because the scaladoc of the `ConfigRepository` trait states the following `@return a copy of the configuration for the given resource`. It seems that the two main implementations respect this however `MockConfigRepository` does not. Therefore, I tend to agree with you.

Reply:
Oh, interesting, so it does specify that a copy is returned. I had missed that. Since the mock config repository does not implement it correctly, as you found, I think it's better to handle this via a separate PR (it's possible that unrelated tests may fail as a result and may need fixing).

Reply:
Regarding unmodifiable `Properties`, there is no built-in way. My concern was something like the `MockConfigRepository` issue you found where copies are not done with the assumption that the caller should not mutate (even though there is no simple way to prevent that - another reason not to use `Properties`).